### ISSUE TITLE
matmul numerical mismatch

### ISSUE THREAD

**Issue by alyashev** [tenstorrent/tt-metal#24872]
### Describe the bug

Here is the ran of the script attached
it shows difference in ulp with the golden pythorch implementation.
The high difference result in a the deverges of the most models we've run.
 
2025-07-10 19:04:35.633 | info     |   SiliconDriver | Opened PCI device 0; KMD version: 1.34.0; API: 2; IOMMU: disabled (pci_device.cpp:198)
2025-07-10 19:04:35.634 | info     |   SiliconDriver | Opened PCI device 0; KMD version: 1.34.0; API: 2; IOMMU: disabled (pci_device.cpp:198)
2025-07-10 19:04:35.641 | info     |          Device | Opening user mode device driver (tt_cluster.cpp:190)
2025-07-10 19:04:35.641 | info     |   SiliconDriver | Opened PCI device 0; KMD version: 1.34.0; API: 2; IOMMU: disabled (pci_device.cpp:198)
2025-07-10 19:04:35.642 | info     |   SiliconDriver | Opened PCI device 0; KMD version: 1.34.0; API: 2; IOMMU: disabled (pci_device.cpp:198)
2025-07-10 19:04:35.649 | info     |   SiliconDriver | Opened PCI device 0; KMD version: 1.34.0; API: 2; IOMMU: disabled (pci_device.cpp:198)
2025-07-10 19:04:35.650 | info     |   SiliconDriver | Opened PCI device 0; KMD version: 1.34.0; API: 2; IOMMU: disabled (pci_device.cpp:198)
2025-07-10 19:04:35.657 | info     |   SiliconDriver | Harvesting mask for chip 0 is 0x201 (NOC0: 0x201, simulated harvesting mask: 0x0). (cluster.cpp:282)
2025-07-10 19:04:35.662 | info     |   SiliconDriver | Opened PCI device 0; KMD version: 1.34.0; API: 2; IOMMU: disabled (pci_device.cpp:198)
2025-07-10 19:04:35.668 | info     |   SiliconDriver | Harvesting mask for chip 1 is 0x202 (NOC0: 0x202, simulated harvesting mask: 0x0). (cluster.cpp:282)
2025-07-10 19:04:35.674 | info     |   SiliconDriver | Opening local chip ids/pci ids: {0}/[0] and remote chip ids {1} (cluster.cpp:147)
2025-07-10 19:04:35.675 | info     |   SiliconDriver | Software version 6.0.0, Ethernet FW version 6.14.0 (Device 0) (cluster.cpp:1039)
2025-07-10 19:04:35.676 | info     |   SiliconDriver | Software version 6.0.0, Ethernet FW version 6.14.0 (Device 1) (cluster.cpp:1039)
2025-07-10 19:04:35.687 | info     |           Metal | AI CLK for device 0 is:   1000 MHz (metal_context.cpp:128)
2025-07-10 19:04:35.975 | info     |           Metal | AI CLK for device 1 is:   1000 MHz (metal_context.cpp:128)
2025-07-10 19:04:36.026 | info     |           Metal | Initializing device 0. Program cache is enabled (device.cpp:428)
range: (-inf, inf) (randn)
mat1:  torch.Size([1024, 1024])  mat2:  torch.Size([1024, 1024])
ulp(2114912256.0000) ulp_mean(4669205.0000)
range: (0, 1) (rand)
mat1:  torch.Size([1024, 1024])  mat2:  torch.Size([1024, 1024])
ulp(327680.0000) ulp_mean(45690.4375)
mat1:  torch.Size([16384, 256])  mat2:  torch.Size([256, 32])
ulp(262144.0000) ulp_mean(44101.5000)
2025-07-10 19:04:37.781 | info     |           Metal | Closing mesh device 1 (mesh_device.cpp:488)
2025-07-10 19:04:37.782 | info     |           Metal | Closing mesh device 0 (mesh_device.cpp:488)
2025-07-10 19:04:37.782 | info     |           Metal | Closing device 0 (device.cpp:468)
2025-07-10 19:04:37.782 | info     |           Metal | Disabling and clearing program cache on device 0 (device.cpp:780)
2025-07-10 19:04:37.830 | info     |           Metal | Closing mesh device 1 (mesh_device.cpp:488)

### Steps to reproduce the issue

1. run the script attached

[test_ttnn_mm_1_1.txt](https://github.com/user-attachments/files/21170253/test_ttnn_mm_1_1.txt)

**Comment by bbradelTT** [comment]
You would need to change everything to float32, including the from_torch calls:
```
    a = ttnn.from_torch(self, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.float32)
    b = ttnn.from_torch(mat2, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.float32)
```
to get higher accuracy.

Could you please try that and see if the accuracy is sufficient?

**Comment by alyashev** [comment]
Thank you for the prompt response.
Here is an update script and run. 
The switch helps a bit but not too mach, and, the major point, we cannot control what precision the model is asking for.
range: (-inf, inf) (randn)
mat1:  torch.Size([1024, 1024]) torch.float32  mat2:  torch.Size([1024, 1024]) torch.float32
ulp(2091490432.0000) ulp_mean(2396315.2500)
range: (0, 1) (rand)
mat1:  torch.Size([1024, 1024]) torch.float32  mat2:  torch.Size([1024, 1024]) torch.float32
ulp(144599.0000) ulp_mean(77221.1094)
mat1:  torch.Size([16384, 256]) torch.float32  mat2:  torch.Size([256, 32]) torch.float32
ulp(36293.0000) ulp_mean(18444.2031)
range: (-inf, inf) (randn)
mat1:  torch.Size([1024, 1024]) torch.bfloat16  mat2:  torch.Size([1024, 1024]) torch.bfloat16
ulp(2109734912.0000) ulp_mean(4047489.2500)
range: (0, 1) (rand)
mat1:  torch.Size([1024, 1024]) torch.bfloat16  mat2:  torch.Size([1024, 1024]) torch.bfloat16
ulp(458752.0000) ulp_mean(56185.0000)
mat1:  torch.Size([16384, 256]) torch.bfloat16  mat2:  torch.Size([256, 32]) torch.bfloat16
ulp(327680.0000) ulp_mean(46232.5000)

[test_ttnn_mm_1_1.txt](https://github.com/user-attachments/files/21171229/test_ttnn_mm_1_1.txt)

**Comment by bbradelTT** [comment]
The ulp values for -inf, inf is probably dominated by some outliers at or close to inf.
327680.0000 / 36293.0000 ~= 9
458752.0000 / 144599.0000 ~= 3
That's a 3-9x improvement.

What type of accuracy are you looking for?

Also, what are you able to control?

Would you be able to set the dtype in the from_torch calls via 
```
    a = ttnn.from_torch(self, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.float32)
    b = ttnn.from_torch(mat2, layout=ttnn.TILE_LAYOUT, device=device, dtype=ttnn.float32)
```
or parameters to matmul (e.g. pass in a compute kernel config and output data type)?

**Comment by alyashev** [comment]
@titash-openxpu could you help as well

**Comment by alyashev** [comment]
we can control both compute kernel config and output data type. if you give me an example I can try.

**Comment by titash-openxpu** [comment]
Attached the ulp mismatch data for 3 different inputs:

input in range (0,1) ->
ulp(393216.0000) ulp_mean(55983.1875)

input in range (-1, 1) ->
ulp(2082340864.0000) ulp_mean(3413410.5000)

input in range (-inf, inf) ->
ulp(2109734912.0000) ulp_mean(4047489.5000)

[mm_mismatch.txt](https://github.com/user-attachments/files/21172391/mm_mismatch.txt)

**Comment by bbradelTT** [comment]
The usual way we adjust precision is to specify the compute kernel config, and you can specify the output data type via the dtype parameter. The code would look something like
```
    compute_kernel_config = ttnn.init_device_compute_kernel_config(
            device.arch(),
            math_fidelity=ttnn.MathFidelity.HiFi4,
            math_approx_mode=False,
            fp32_dest_acc_en=True,
            packer_l1_acc=False,
        )
    dev = ttnn.matmul(a, b, dtype=ttnn.float32, compute_kernel_config=compute_kernel_config)
```

I'd expect the ulp to increase more. I need to look at why the change is what it is.


Also, what type of accuracy are you looking for?

**Comment by alyashev** [comment]
We would like too keep HiFi version on par with pytorch precision in 2ulp bound.
Is it achievable?

**Comment by alyashev** [comment]
after inserting HiPres and fixing upl calculation the matmul on TT (fp32/bfloat16) looks like this
range: (-inf, inf) (randn)
mat1:  torch.Size([1024, 1024]) torch.float32  mat2:  torch.Size([1024, 1024]) torch.float32
ulp(2087748992.0000) ulp_mean(2415242.5000)
range: (0, 1) (rand)
mat1:  torch.Size([1024, 1024]) torch.float32  mat2:  torch.Size([1024, 1024]) torch.float32
ulp(152171.0000) ulp_mean(76966.5781)
mat1:  torch.Size([16384, 256]) torch.float32  mat2:  torch.Size([256, 32]) torch.float32
ulp(35994.0000) ulp_mean(17557.8457)
range: (-inf, inf) (randn)
mat1:  torch.Size([1024, 1024]) torch.bfloat16  mat2:  torch.Size([1024, 1024]) torch.bfloat16
ulp(31902.0000) ulp_mean(35.8275)
range: (0, 1) (rand)
mat1:  torch.Size([1024, 1024]) torch.bfloat16  mat2:  torch.Size([1024, 1024]) torch.bfloat16
ulp(2.0000) ulp_mean(1.0515)
mat1:  torch.Size([16384, 256]) torch.bfloat16  mat2:  torch.Size([256, 32]) torch.bfloat16
ulp(1.0000) ulp_mean(0.0840)

we'r interested in fp32 and wider range in the 2 ulp bound

[util.txt](https://github.com/user-attachments/files/21189991/util.txt)

[test_ttnn_mm_1_1.txt](https://github.com/user-attachments/files/21189994/test_ttnn_mm_1_1.txt)

**Comment by alyashev** [comment]
updated the script.

[util.txt](https://github.com/user-attachments/files/21201418/util.txt)

[test_ttnn_mm_1_1.txt](https://github.com/user-attachments/files/21201420/test_ttnn_mm_1_1.txt)

**Comment by bbradelTT** [comment]
@alyashev 
Please change the packer parameter to true:
```
            packer_l1_acc=True,
```

With the updated script, except where torch device is "cpu", I get
```
collecting ... range: (-inf, inf) (randn)
mat1:  torch.Size([1024, 1024]) torch.float32  mat2:  torch.Size([1024, 1024]) torch.float32
ulp(2045766656.0000) ulp_mean(408205.5625)
range: (0, 1) (rand)
mat1:  torch.Size([1024, 1024]) torch.float32  mat2:  torch.Size([1024, 1024]) torch.float32
ulp(24570.0000) ulp_mean(14525.7500)
mat1:  torch.Size([16384, 256]) torch.float32  mat2:  torch.Size([256, 32]) torch.float32
ulp(25260.0000) ulp_mean(14015.4229)
range: (-inf, inf) (randn)
mat1:  torch.Size([1024, 1024]) torch.bfloat16  mat2:  torch.Size([1024, 1024]) torch.bfloat16
ulp(31535.0000) ulp_mean(10.4623)
range: (0, 1) (rand)
mat1:  torch.Size([1024, 1024]) torch.bfloat16  mat2:  torch.Size([1024, 1024]) torch.bfloat16
ulp(1.0000) ulp_mean(0.3938)
mat1:  torch.Size([16384, 256]) torch.bfloat16  mat2:  torch.Size([256, 32]) torch.bfloat16
ulp(1.0000) ulp_mean(0.3826)
```

Even HiFi2 is within 2 ulp for the last two entries:
```
    compute_kernel_config = ttnn.init_device_compute_kernel_config(
        device.arch(),
        math_fidelity=ttnn.MathFidelity.LoFi,
            math_approx_mode=False,
            fp32_dest_acc_en=True,
            packer_l1_acc=True,
    )
```
produces
```
mat1:  torch.Size([1024, 1024]) torch.bfloat16  mat2:  torch.Size([1024, 1024]) torch.bfloat16
ulp(2.0000) ulp_mean(0.8022)
mat1:  torch.Size([16384, 256]) torch.bfloat16  mat2:  torch.Size([256, 32]) torch.bfloat16
ulp(2.0000) ulp_mean(0.7719)
```

inf may cause issues for ulp.

Our matrix unit does operations on TF32 inputs (https://github.com/tenstorrent/tt-isa-documentation/blob/main/WormholeB0/TensixTile/TensixCoprocessor/MatrixUnit.md).
Just the accumulation can be done on FP32.

We will not be able to get fp32 in the 2 ulp bound when inputs are truncated to tf32.

**Comment by bbradelTT** [comment]
@alyashev we are getting to 1-2ulp for bfloat16. Is that sufficient to close this issue?

**Comment by alyashev** [comment]
I'll check it out shortly, sorry for the delay.

**Comment by alyashev** [comment]
Let me parphrase what you saying:
1. matmul cannot reach 2ulp bound in any format in  (-inf, inf) range.
2. matmul fp32 cannot guarantee 2ulp in (0,1) range.
am I right?

**Comment by alyashev** [comment]
I've rerun this:
# Copyright (c) 2025 OpenXPU Inc

import torch
import ttnn
import numpy as np

from util import calculate_ulp_max

def aten__mm__default_test(device, self, mat2, dtype_default = torch.bfloat16, seed = 404):
    if seed is not None:
        torch.manual_seed(seed)
    print("mat1: ", self.shape, self.dtype, " mat2: ", mat2.shape, mat2.dtype)
    ref = torch.matmul(self,mat2)
    a = ttnn.from_torch(self, layout=ttnn.TILE_LAYOUT, device=device)
    b = ttnn.from_torch(mat2, layout=ttnn.TILE_LAYOUT, device=device)
    compute_kernel_config = ttnn.init_device_compute_kernel_config(
        device.arch(),
        math_fidelity=ttnn.MathFidelity.HiFi4,
            math_approx_mode=False,
            fp32_dest_acc_en=True,
            packer_l1_acc=False,
    )
    dev = ttnn.matmul(a, b, compute_kernel_config=compute_kernel_config)
    dev = dev.cpu().to_torch()
    ulp_max, ulp_mean = calculate_ulp_max(dev, ref)
    print(f'ulp({ulp_max:.4f}) ulp_mean({ulp_mean:.4f})')


device_id = 0
device = ttnn.open_device(device_id=device_id)

dtype_default = torch.float32

m001 = torch.randn([1024,1024], dtype=dtype_default, device='cpu')
m002 = torch.randn([1024,1024], dtype=dtype_default, device='cpu')
m003 = torch.rand([1024,1024], dtype=dtype_default, device='cpu')
m004 = torch.rand([1024,1024], dtype=dtype_default, device='cpu')
m005 = torch.rand([16384,256], dtype=dtype_default, device='cpu')
m006 = torch.rand([256,32], dtype=dtype_default, device='cpu')


print("range: (-inf, inf) (randn)")
aten__mm__default_test(device, m001, m002)
print("range: (0, 1) (rand)")
aten__mm__default_test(device, m003, m004)
aten__mm__default_test(device, m005, m006)



dtype_default = torch.bfloat16

m001 = torch.randn([1024,1024], dtype=dtype_default, device='cpu')
m002 = torch.randn([1024,1024], dtype=dtype_default, device='cpu')
m003 = torch.rand([1024,1024], dtype=dtype_default, device='cpu')
m004 = torch.rand([1024,1024], dtype=dtype_default, device='cpu')
m005 = torch.rand([16384,256], dtype=dtype_default, device='cpu')
m006 = torch.rand([256,32], dtype=dtype_default, device='cpu')

print("range: (-inf, inf) (randn)")
aten__mm__default_test(device, m001, m002)
print("range: (0, 1) (rand)")
aten__mm__default_test(device, m003, m004)
aten__mm__default_test(device, m005, m006)

ttnn.close_device(device)

**Comment by alyashev** [comment]
and still got this:
range: (-inf, inf) (randn)
mat1:  torch.Size([1024, 1024]) torch.float32  mat2:  torch.Size([1024, 1024]) torch.float32
ulp(2091514368.0000) ulp_mean(2384016.2500)
range: (0, 1) (rand)
mat1:  torch.Size([1024, 1024]) torch.float32  mat2:  torch.Size([1024, 1024]) torch.float32
ulp(144174.0000) ulp_mean(77200.3438)
mat1:  torch.Size([16384, 256]) torch.float32  mat2:  torch.Size([256, 32]) torch.float32
ulp(36763.0000) ulp_mean(17603.8027)
range: (-inf, inf) (randn)
mat1:  torch.Size([1024, 1024]) torch.bfloat16  mat2:  torch.Size([1024, 1024]) torch.bfloat16
ulp(31902.0000) ulp_mean(35.8275)
range: (0, 1) (rand)
mat1:  torch.Size([1024, 1024]) torch.bfloat16  mat2:  torch.Size([1024, 1024]) torch.bfloat16
ulp(2.0000) ulp_mean(1.0515)
mat1:  torch.Size([16384, 256]) torch.bfloat16  mat2:  torch.Size([256, 32]) torch.bfloat16
ulp(1.0000) ulp_mean(0.0840)

is it what expected?

**Comment by bbradelTT** [comment]
> Let me parphrase what you saying:
> 
> 1. matmul cannot reach 2ulp bound in any format in  (-inf, inf) range.
> 2. matmul fp32 cannot guarantee 2ulp in (0,1) range.
>    am I right?

Yes, that's correct.


For the results that you got when you re-ran, yes, that is expected.

If you switch to
```
compute_kernel_config = ttnn.init_device_compute_kernel_config(
device.arch(),
math_fidelity=ttnn.MathFidelity.HiFi4,
math_approx_mode=False,
fp32_dest_acc_en=True,
packer_l1_acc=True,
)
```
the ulp values will be slightly better.

**Comment by bbradelTT** [comment]
@alyashev is there any more work for this issue, or can we close it?

**Comment by alyashev** [comment]
Pls close it. Thank you a lot.
